src/objective_types/oncedifferentiable.jl

```diff
@@ -0,0 +1,40 @@
+# Used for objectives and solvers where the gradient is available/exists
+mutable struct OnceDifferentiable{TF, TDF, TX, Tcplx<:Union{Val{true},Val{false}}} <: AbstractObjective
+    f # objective
+    df # (partial) derivative of objective
+    fdf # objective and (partial) derivative of objective
+    F::TF # cache for f output
```

Comment:
should `f`, `df`, `fdf` have type parameters? Otherwise this causes a type instability?

Reply:
same for `TwiceDifferentiable`

Reply:
They could, but they weren't. It will require recompilation for each new instance which might be an annoyance if people create many anonymous functions. We could allow for both though.

Reply:
Personally, I don't care - my objectives are very expensive. This was more targeted towards the use-case of many cheap optimisations. But you are right, if one has to recompile each time then this is not helpful.

Reply:
Exactly. It's fine to compile for say each input type (Float64, Vector ...) but I do get the impression that a lot of people are iteratively/repeatedly creating anonymous functions that they minimize in fractions of seconds (but maybe thousands of times), and this would seriously hurt them. As I said, we can have a special type for the non-default type if necessary.

Anyway, the fact that you cannot infer the output type isn't really too bad, as we always know exactly what the type of the cache variable is. This means that even if the function ends up spitting out Float32's, it'll automatically be promoted to Float64's, and the majority of the code will be compiled under the assumption that it is indeed Float64s.